daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Tagged.scala

```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.data
+
+import scala.language.higherKinds
+
+abstract class Tagged[X] {
+  type T <: X
+  def apply(t: X): T
+  def subst[F[_]](fx: F[X]): F[T]
+}
+
+object Tagged {
+  def apply[X]: Tagged[X] = new Tagged[X] {
+    override type T = X
+    override def apply(x: X): T = x
+    override def subst[F[_]](fx: F[X]): F[T] = fx
+  }
+}
```

Comment:
I discourage anyone from thinking there is any one-size-fits-all approach to tagging. Capturing common patterns is valuable, but it would be unwise to attempt to fit all such newtypes into one library's pattern. I briefly discuss some of that here https://youtu.be/ej-_FUZClWc?t=2314 (38:34 to 39:17)

With respect to allowing subtyping there are two major caveats, which I talk about earlier in the same talk https://youtu.be/ej-_FUZClWc?t=1587 (26:27 to 30:01). If neither of these apply to a particular use case then it is fine to use subtyping. Note however that both caveats take the form of “you can’t decide later that a particular feature should be added”.

For inspiration you have taken the form most similar to the "bespoke" style that we use for most newtypes in the daml-lf codebase: the one where the tag is the singleton type of the containing structure. There are actually two other ways to define a general-purpose subtype tag.

```scala
// one is closer to @@; the only change is to add the upper bound to the visible signature
type <@[+X, T] <: X

// the other is to use scala's `with` instead of a custom infix tag attacher
type T  // define as type T = Any in the concrete structure
def apply(x: X): X with T  // = x in the concrete structure
def subst[F[_]](fx: F[X]): F[X with T] // likewise
```

As in Scalaz you only have one structure instead of one-per-tagged-type for these.

Finally I suggest a name that suggests the properties of this particular tag, like `SubtypeTagged` or `TranslucentTagged`.

Reply:
Thanks a lot for the explanation. 
I went for the subtype approach, because a least on of the "caveats" your are speaking about is a property I am in fact looking for. 
I want that any construction (from scratch or using mapping operation) of new Submitted/Committed transactions are explicitly tagged each time.